src/main/java/org/icatproject/core/manager/search/OpensearchQueryBuilder.java

```diff
@@ -0,0 +1,213 @@
+     * @return <code>{"term": {`field`: `value`}}</code>
+     */
+    public static JsonObject buildTermQuery(String field, double value) {
+        return Json.createObjectBuilder().add("term", Json.createObjectBuilder().add(field, value)).build();
+    }
+
+    /**
+     * @param field Field containing on of the terms.
+     * @param values JsonArray of possible terms.
+     * @return <code>{"terms": {`field`: `values`}}</code>
+     */
+    public static JsonObject buildTermsQuery(String field, JsonArray values) {
+        return Json.createObjectBuilder().add("terms", Json.createObjectBuilder().add(field, values)).build();
+    }
+
+    /**
+     * @param field Field to apply the range to.
+     * @param lowerValue Lowest allowed value in the range.
+     * @param upperValue Highest allowed value in the range.
+     * @return <code>{"range": {`field`: {"gte": `upperValue`, "lte": `lowerValue`}}}</code>
+     */
+    public static JsonObject buildDoubleRangeQuery(String field, Double lowerValue, Double upperValue) {
+        JsonObjectBuilder fieldBuilder = Json.createObjectBuilder();
+        if (lowerValue != null) fieldBuilder.add("gte", lowerValue);
+        if (upperValue != null) fieldBuilder.add("lte", upperValue);
+        JsonObjectBuilder rangeBuilder = Json.createObjectBuilder().add(field, fieldBuilder);
+        return Json.createObjectBuilder().add("range", rangeBuilder).build();
+    }
+
+    /**
+     * @param field Field to apply the range to.
+     * @param lowerValue Lowest allowed value in the range.
+     * @param upperValue Highest allowed value in the range.
+     * @return <code>{"range": {`field`: {"gte": `upperValue`, "lte": `lowerValue`}}}</code>
+     */
+    public static JsonObject buildLongRangeQuery(String field, Long lowerValue, Long upperValue) {
+        JsonObjectBuilder fieldBuilder = Json.createObjectBuilder();
+        if (lowerValue != null) fieldBuilder.add("gte", lowerValue);
+        if (upperValue != null) fieldBuilder.add("lte", upperValue);
+        JsonObjectBuilder rangeBuilder = Json.createObjectBuilder().add(field, fieldBuilder);
+        return Json.createObjectBuilder().add("range", rangeBuilder).build();
+    }
+
+    /**
+     * @param field Field to apply the range to.
+     * @param lowerValue Lowest allowed value in the range.
+     * @param upperValue Highest allowed value in the range.
+     * @return <code>{"range": {`field`: {"gte": `upperValue`, "lte": `lowerValue`}}}</code>
+     */
+    public static JsonObject buildRangeQuery(String field, JsonNumber lowerValue, JsonNumber upperValue) {
+        JsonObjectBuilder fieldBuilder = Json.createObjectBuilder();
+        if (lowerValue != null) fieldBuilder.add("gte", lowerValue);
+        if (upperValue != null) fieldBuilder.add("lte", upperValue);
+        JsonObjectBuilder rangeBuilder = Json.createObjectBuilder().add(field, fieldBuilder);
+        return Json.createObjectBuilder().add("range", rangeBuilder).build();
+    }
```

Comment:
I don't have experience with generics, but I don't see a way of reducing the duplication (at least in terms of the arguments). The [`add` methods on `JsonObjectBuilder`](https://docs.oracle.com/javaee/7/api/javax/json/JsonObjectBuilder.html) expect either `long`, `double` or `JsonValue` (in addition to other non-relevant types). As I understand it, we would need `add` to accept a class that all of those extend, but there isn't one? Or am I missing something?

What we can do is refactor the last two lines of all of them.

Reply:
> As I understand it, we would need `add` to accept a class that all of those extend

I don't think so. The types of `lowerValue` and `upperValue` would be the same as those provided to `buildRangeQuery` - i.e. if you pass a `Double` for the `lowerValue` arg then its type will be a `Double`, see the snippet below.

```Java
public static <T> void buildRangeQuery(String field, T lowerValue, T upperValue) {
  System.out.println(lowerValue.getClass());
  System.out.println(upperValue.getClass());
}
buildRangeQuery("", 1, 1);
buildRangeQuery("", 1.0, 1.0);
buildRangeQuery("", 1L, 1L);
buildRangeQuery("", "ab", "ab");

Output:
>>>class java.lang.Integer
>>>class java.lang.Integer
>>>class java.lang.Double
>>>class java.lang.Double
>>>class java.lang.Long
>>>class java.lang.Long
>>>class java.lang.String
>>>class java.lang.String
```

So what I was thinking is replacing the three methods with a generics method as the only difference in these methods is the argument types. However, if you do not have experience with generics then it might be best to leave the code as it is because while I may think that this would work, I don't want to find out that it causes some issues, which I haven't thought of, further down the line.

Reply:
Oh, I see. I thought it would just pass the actual type to `JsonObjectBuilder.add` instead of `T` but clearly not. I do not know of a way around this, sorry. Best to leave the code as it is.